src/redux/todosSlice.ts

```diff
@@ -0,0 +1,73 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+export interface Todo {
+  id: string;
+  text: string;
+  completed: boolean;
+}
+
+interface TodosState {
+  todos: Todo[];
+}
+
+const saveState = (state: Todo[]) => {
+  try {
+    const serializedState = JSON.stringify(state);
+    localStorage.setItem('todos', serializedState);
+  } catch (error) {
+    console.error('Error saving state:', error);
+  }
+};
+
+const loadState = (): TodosState => {
+  const emptyState = {
+    todos: [],
+  };
+  try {
+    const serializedState = localStorage.getItem('todos');
+    if (serializedState === null) {
+      return emptyState;
+    }
+    return {
+      todos: JSON.parse(serializedState),
+    };
+  } catch (error) {
+    console.error('Error loading state:', error);
+    return emptyState;
+  }
+};
+
+const initialState: TodosState = loadState();
+
+const todosSlice = createSlice({
+  name: 'todosList',
+  initialState,
+  reducers: {
+    addTodo: (state, action: PayloadAction<string>) => {
+      const newTodo: Todo = {
+        id: Date.now().toString(),
+        text: action.payload,
+        completed: false,
+      };
+      state.todos.push(newTodo);
```

Comment:
Here you are mutating the source array, which is heavily discouraged (unless you are using something like `immer`). You need to create a new array with your todo in it

Reply:
Redux Toolkit's createReducer API uses Immer internally automatically. So, it's already safe to "mutate" the state inside of any case reducer function that is passed to createReducer or createSlice. Please check the link below:
https://redux-toolkit.js.org/usage/immer-reducers#redux-toolkit-and-immer